### PR TITLE
Fix tests setup issues

### PR DESCRIPTION
**Pull request by medikoo**
## What did you implement:

Addresses #6116 (not closing yet as there are some further steps to make)

- Improve `setTimeout` stub in `lib/plugins/aws/provider/awsProvider.test.js`, setting it to _sync_, broke some core utilities (as bluebird, which in effect reported unhandled rejections immediatally after rejected promise was created)
- Ensure mangling of `process.env` vars does not introduce side effects (so `process.env` vars are left in same shape as it was pior test run), and that existing env vars do not influence test output where it's not expected.
- Introduce alternative `test-isolated` test runner, which runs each test file in individual process.
It's the way to confirm that tests do no work by chance of side effects introduced by other tests, and that they properly cleanup state they've introduced.
- Fixed issues discovered via isolated test runs (missing tooling setup in most cases)
- Ensure tests do not  wait for `stdin` input (some tests hanged if the process didn't report a TTY for stdin)
- Clean`mocha` setup:
  - Reordered CLI args so options appear prior tests path (it's only when mocha signals pass of invalid options)
  - Remove `--no-exit` option, as it's how Mocha behaves by default [since v4](https://github.com/mochajs/mocha/pull/3034)
  - Remove not effective options:
    - `--recursive` is not needed, as glob pattern we use already forces recursive check
    - `-R spec`, `spec` is the default reporter, no need to explicitly set it
- Improve .gitignore rules (`/bin` folder was ignored, and that's not right)

__Note:__ It'll also be good to add `npm run test-isolated` to CI, so it's run at least once for chosen Node.js version (to confirm we don't have regression in this context).
Still first I would improve `.travis.yml` to be configured with [stages](https://docs.travis-ci.com/user/build-stages), therefore I'd rather leave it out of scope of this PR.

### CODE REVIEW

File: bin/test-isolated [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  );
+
+const initialGitStatusDeferred = resolveGitStatus();
+
+const run = path => {
+  const onFinally = () =>
+    Promise.all([initialGitStatusDeferred, resolveGitStatus()]).then(
+      ([initialStatus, currentStatus]) => {
+        if (initialStatus !== currentStatus) {
+          process.stderr.write(chalk.red.bold(`${path} didn't clean creted temporary files\n\n`));
```
Comment (pmuens) [inline bin/test-isolated]: There might be a type here.

File: bin/test-isolated [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+      process.stderr.write(error.stderrBuffer);
+      throw error;
+    }
+  );
+
+const initialGitStatusDeferred = resolveGitStatus();
+
+const run = path => {
+  const onFinally = () =>
+    Promise.all([initialGitStatusDeferred, resolveGitStatus()]).then(
```
Comment (pmuens) [inline bin/test-isolated]: What's the difference between the two Promises here? I understand what this code should do but aren't they technically the same?

File: lib/classes/Error.test.js [hunk 3]
```diff
@@ -67,9 +67,15 @@ describe('Error', () => {
...
   describe('#logError()', () => {
+    let BK_SLS_DEBUG;
     beforeEach(() => {
+      BK_SLS_DEBUG = process.env.SLS_DEBUG;
       delete process.env.SLS_DEBUG;
     });
+    afterEach(() => {
+      if (BK_SLS_DEBUG) process.env.SLS_DEBUG = BK_SLS_DEBUG;
+      else delete process.env.SLS_DEBUG;
+    });
```
Comment (pmuens) [inline lib/classes/Error.test.js]: Can we add some `\n` after `BK_SLS_DEBUG` and `beforeEach`?

Out of curiosity: What does `BK` stand for?

File: lib/classes/PluginManager.test.js [hunk 4]
```diff
@@ -1438,7 +1443,9 @@ describe('PluginManager', () => {
       return pluginManager.run(commandsArray).then(() => {
         expect(consoleLogStub.called).is.equal(true);
         pluginManager.serverless.cli.log.restore();
-        process.env.SLS_DEBUG = undefined;
+      }).finally(() => {
```
Comment (pmuens) [inline lib/classes/PluginManager.test.js]: Is this the preferred way to do cleanup? I always love `afterEach` or `afterAll` becaus this keeps it consistent and makes it easier to understand the setup / teardown process.

File: lib/classes/PluginManager.test.js [hunk 5]
```diff
@@ -1727,7 +1735,9 @@ describe('PluginManager', () => {
...
```
Comment (pmuens) [inline lib/classes/PluginManager.test.js]: Same question applies here.

File: lib/plugins/aws/invokeLocal/index.test.js [hunk 6]
```diff
@@ -250,7 +260,9 @@ describe('AwsInvokeLocal', () => {
   });
 
   describe('#loadEnvVars()', () => {
+    let BK_AWS_PROFILE;
```
Comment (pmuens) [inline lib/plugins/aws/invokeLocal/index.test.js]: Can we add a newline here?

File: lib/plugins/aws/invokeLocal/index.test.js [hunk 7]
```diff
@@ -83,6 +83,7 @@ describe('AwsInvokeLocal', () => {
   });
 
   describe('#extendedValidate()', () => {
+    let backupIsTTY;
```
Comment (pmuens) [inline lib/plugins/aws/invokeLocal/index.test.js]: Can we add a newline here?

File: lib/plugins/aws/invoke/index.test.js [hunk 8]
```diff
@@ -51,6 +51,7 @@ describe('AwsInvoke', () => {
   });
 
   describe('#extendedValidate()', () => {
+    let backupIsTTY;
```
Comment (pmuens) [inline lib/plugins/aws/invoke/index.test.js]: Can we add a newline here?

File: lib/plugins/aws/invokeLocal/index.test.js [hunk 9]
```diff
@@ -1183,6 +1196,7 @@ describe('AwsInvokeLocal', () => {
       delete require.cache[require.resolve('./index')];
       delete require.cache[require.resolve('child_process')];
       serverless.pluginManager.spawn.restore();
+      return fse.remove('.serverless');
```
Comment (pmuens) [inline lib/plugins/aws/invokeLocal/index.test.js]: Should we move this to a `afterEach` / `afterAll` block?

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 10]
```diff
@@ -41,10 +41,16 @@ describe('AwsProvider', () => {
...
   describe('#constructor()', () => {
-    afterEach('Environment Variable Cleanup', () => {
-      // clear env
+    let backupProxyEnv;
+
+    beforeEach('Environment Variable Cleanup', () => {
+      backupProxyEnv = process.env.proxy;
       delete process.env.proxy;
     });
+    afterEach('Environment Variable Cleanup', () => {
```
Comment (pmuens) [inline lib/plugins/aws/provider/awsProvider.test.js]: Can we add a newline before the `afterEach` block here?

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 11]
```diff
@@ -59,18 +65,20 @@ describe('AwsProvider', () => {
...
-      const newAwsProvider = new AwsProvider(serverless, options);
-
-      expect(typeof newAwsProvider.sdk.config.logger).to.not.equal('undefined');
+      try {
+        const newAwsProvider = new AwsProvider(serverless, options);
 
-      // reset env
-      process.env.SLS_DEBUG = BK_SLS_DEBUG;
+        expect(typeof newAwsProvider.sdk.config.logger).to.not.equal('undefined');
+      } finally {
```
Comment (pmuens) [inline lib/plugins/aws/provider/awsProvider.test.js]: same question about the cleanup process.

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 12]
```diff
@@ -81,13 +89,16 @@ describe('AwsProvider', () => {
...
-      const newAwsProvider = new AwsProvider(serverless, options);
-
-      expect(typeof newAwsProvider.sdk.config.httpOptions.timeout).to.not.equal('undefined');
+      try {
+        const newAwsProvider = new AwsProvider(serverless, options);
 
-      // clear env
-      delete process.env.AWS_CLIENT_TIMEOUT;
+        expect(typeof newAwsProvider.sdk.config.httpOptions.timeout).to.not.equal('undefined');
+      } finally {
```
Comment (pmuens) [inline lib/plugins/aws/provider/awsProvider.test.js]: Same question about the cleanup process.

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 13]
```diff
@@ -131,10 +142,16 @@ describe('AwsProvider', () => {
...
     describe('certificate authority - environment variable', () => {
-      afterEach('Environment Variable Cleanup', () => {
-        // clear env
+      let backupCaEnv;
+
+      beforeEach('Environment Variable Cleanup', () => {
+        backupCaEnv = process.env.ca;
         delete process.env.ca;
       });
+      afterEach('Environment Variable Cleanup', () => {
```
Comment (pmuens) [inline lib/plugins/aws/provider/awsProvider.test.js]: Can we add a newline here? 😬

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 14]
```diff
@@ -131,10 +142,16 @@ describe('AwsProvider', () => {
...
+        if (backupCaEnv) process.env.ca = backupCaEnv;
+        else delete process.env.ca;
+      });
```
Comment (pmuens) [inline lib/plugins/aws/provider/awsProvider.test.js]: And here as well?

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 15]
```diff
@@ -165,14 +182,25 @@ describe('AwsProvider', () => {
       const tmpdir = os.tmpdir();
       let file1 = null;
       let file2 = null;
+      let backupEnv;
```
Comment (pmuens) [inline lib/plugins/aws/provider/awsProvider.test.js]: Can we add a newline here?

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 16]
```diff
@@ -1011,13 +1046,25 @@ describe('AwsProvider', () => {
...
-      const credentials = newAwsProvider.getCredentials();
-      expect(credentials.credentials.accessKeyId).to.equal(testVal.accessKeyId);
-      expect(credentials.credentials.secretAccessKey).to.equal(testVal.secretAccessKey);
-      expect(credentials.credentials.sessionToken).to.equal(testVal.sessionToken);
+      try {
+        const credentials = newAwsProvider.getCredentials();
+        expect(credentials.credentials.accessKeyId).to.equal(testVal.accessKeyId);
+        expect(credentials.credentials.secretAccessKey).to.equal(testVal.secretAccessKey);
+        expect(credentials.credentials.sessionToken).to.equal(testVal.sessionToken);
+      } finally {
```
Comment (pmuens) [inline lib/plugins/aws/provider/awsProvider.test.js]: Should we move this in a separate `afterEach` / `afterAll` block?

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 17]
```diff
@@ -1026,52 +1073,94 @@ describe('AwsProvider', () => {
...
```
Comment (pmuens) [inline lib/plugins/aws/provider/awsProvider.test.js]: Same cleanup question applies here.

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 18]
```diff
@@ -1026,52 +1073,94 @@ describe('AwsProvider', () => {
...
 
     it('should get credentials from environment declared for-all-stages profile', () => {
+      const BK_AWS_PROFILE = process.env.AWS_PROFILE;
       process.env.AWS_PROFILE = 'notDefault';
-      const credentials = newAwsProvider.getCredentials();
-      expect(credentials.credentials.profile).to.equal('notDefault');
+      try {
+        const credentials = newAwsProvider.getCredentials();
+        expect(credentials.credentials.profile).to.equal('notDefault');
+      } finally {
```
Comment (pmuens) [inline lib/plugins/aws/provider/awsProvider.test.js]: Same cleanup question applies here.

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 19]
```diff
@@ -1026,52 +1073,94 @@ describe('AwsProvider', () => {
...
 
     it('should get credentials from environment declared stage-specific profile', () => {
+      const BK_AWS_TESTSTAGE_PROFILE = process.env.AWS_TESTSTAGE_PROFILE;
       process.env.AWS_TESTSTAGE_PROFILE = 'notDefault';
-      const credentials = newAwsProvider.getCredentials();
-      expect(credentials.credentials.profile).to.equal('notDefault');
+      try {
+        const credentials = newAwsProvider.getCredentials();
+        expect(credentials.credentials.profile).to.equal('notDefault');
+      } finally {
```
Comment (pmuens) [inline lib/plugins/aws/provider/awsProvider.test.js]: Same cleanup question applies here.

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 20]
```diff
@@ -1026,52 +1073,94 @@ describe('AwsProvider', () => {
...
-      newAwsProvider.options['aws-profile'] = 'notDefault';
-
-      const credentials = newAwsProvider.getCredentials();
-      expect(credentials.credentials.profile).to.equal('notDefault');
+      try {
+        newAwsProvider.options['aws-profile'] = 'notDefault';
+
+        const credentials = newAwsProvider.getCredentials();
+        expect(credentials.credentials.profile).to.equal('notDefault');
+      } finally {
```
Comment (pmuens) [inline lib/plugins/aws/provider/awsProvider.test.js]: Same cleanup question applies here.

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 21]
```diff
@@ -1026,52 +1073,94 @@ describe('AwsProvider', () => {
...
+        newAwsProvider.options['aws-profile'] = 'notDefault';
 
-      serverless.service.provider.profile = 'notDefaultTemporary2';
+        serverless.service.provider.profile = 'notDefaultTemporary2';
 
-      const credentials = newAwsProvider.getCredentials();
-      expect(credentials.credentials.profile).to.equal('notDefault');
+        const credentials = newAwsProvider.getCredentials();
+        expect(credentials.credentials.profile).to.equal('notDefault');
+      } finally {
```
Comment (pmuens) [inline lib/plugins/aws/provider/awsProvider.test.js]: Same cleanup question applies here.

File: lib/plugins/aws/provider/awsProvider.test.js [hunk 22]
```diff
@@ -1026,52 +1073,94 @@ describe('AwsProvider', () => {
...
 
-      serverless.service.provider.profile = 'notDefaultTemporary';
+      try {
+        serverless.service.provider.profile = 'notDefaultTemporary';
 
-      const credentials = newAwsProvider.getCredentials();
-      expect(credentials.credentials.profile).to.equal('notDefault');
+        const credentials = newAwsProvider.getCredentials();
+        expect(credentials.credentials.profile).to.equal('notDefault');
+      } finally {
```
Comment (pmuens) [inline lib/plugins/aws/provider/awsProvider.test.js]: Same cleanup question applies here.

File: package.json [hunk 23]
```diff
@@ -52,6 +52,7 @@
   },
   "scripts": {
     "test-bare": "node bin/test",
+    "test-isolated": "node bin/test-isolated",
```
Comment (pmuens) [inline package.json]: Should this be our default? Is there a need to keep the `test-bare`? / `test`?

File: bin/test-isolated (hunk 2 above)
Comment (medikoo) [reply bin/test-isolated]: First one exposes `git status` as it was _before_ test was run, and second one resolves one _after_ test was run.

If there's a difference between both, it means some temporary files weren't cleaned up

File: lib/classes/Error.test.js (hunk 3 above)
Comment (medikoo) [reply lib/classes/Error.test.js]: I believe it translates to _BACKUP_. I actually reused convention that I already found in test files:
https://github.com/serverless/serverless/blob/fdd8335000c003e3138418c2dc7a261bc6e2fc0a/lib/plugins/aws/provider/awsProvider.test.js#L66

If I was to improve it, I would use utility as: https://github.com/medikoo/process-utils#override-env It'll provide cleanest setup for this kind of mangling. Still it won't work in Node.js v4. so I'm relucant to do it before we officially drop that support with: https://github.com/serverless/serverless/pull/6077

(added `\n`)

File: lib/classes/PluginManager.test.js (hunk 4 above)
Comment (medikoo) [reply lib/classes/PluginManager.test.js]: In that case it's only single test out of 10+ that depends on it, and I'm not sure is it nice to create dedicated describe here, and mangling this env var for other tests might not turn desired.

After we drop Node.js v4 support, I can update it to something as:

```javascript
it('....', overrideEnv(() => {
  process.env.SLS_DEBUG = '*';
  // test logic
}));
```

File: package.json (hunk 23 above)
Comment (medikoo) [reply package.json]: It's slower, but I would definitely run it in one Node.js version run, and I planned to configure that together with migration of Travis.yml to stages (which we agreed to pursue after https://github.com/serverless/serverless/pull/6132 is in).

Concerning `test-bare`. I believe we need to have command where we run tests but with no coverage report. It's faster and doesn't mangle the stack traces.

I would also run coverage report just once in CI, and in all other cases pursue regular test.

We may rename scripts so `test` is just test and `coverage` is test with coverage.
But I would also leave it for .travis.yml conversion PR

**Pull request merged**